avr-hal-generic/src/serial.rs

```diff
@@ -89,6 +89,32 @@ macro_rules! impl_usart {
                     _clock: ::core::marker::PhantomData,
                 }
             }
+		}
+
+		$crate::paste::item! {
+	        impl<CLOCK, RX_MODE> $Usart<CLOCK, RX_MODE>
+	        where
+	            CLOCK: $crate::clock::Clock,
+	            RX_MODE: $crate::port::mode::InputMode,
+	        {
+	            /// Helper method for splitting this read/write object into two halves.
+	            ///
+	            /// The two halves returned implement the `Read` and `Write` traits, respectively.
+	            pub fn split(self) -> ([<Read $Usart>]<CLOCK, RX_MODE>, [<Write $Usart>]<CLOCK>) {
+	                (
+	                    [<Read $Usart>] {
+	                        p: unsafe { ::core::ptr::read(&self.p) },
```

Comment:
I don't really like this, but I'm not sure there's a better way.

Reply:
If I'm not mistaken, this breaks safety.  The reader and write halves are now completely separate so I could move one of the two e.g. into an interrupt context and there would be no synchronization to ensure the two don't interfere.

I see a few ways to fix this:

- Use an async mutex (similar to [this one](https://github.com/ferrous-systems/async-on-embedded/blob/master/async-cortex-m/src/unsync/mutex.rs)) to wrap the peripheral and grant access to multiple tasks.  This would mean, no splitting is needed.
- Use a mutex of sorts internally to synchronize access between reader and writer.  E.g. a plain RefCell which would ensure the halves have to remain in the same hardware thread.
- Synchronize the halves by turning off interrupts in the read and write methods.  I feel like this is not a nice approach though, because it could potentially lock up the system for long times.

Reply:
Although I could mark it as `!Send` to prevent you from sending it across an interrupt bound (we should have that invariant, as otherwise you could do some bad sync stuff), I don't think that's a problem in the first place—reading and writing is done in the same register, but I don't think there's anything that can go wrong. Reading is done in two (main) instructions:

1. Read control register A
2. Read data register

Writing is done in two (main) instructions as well:

1. Read control register A
2. Write data register

But I don't see how placing an interrupt in between either of those could cause problems, as it's not like writing can empty the read buffer or reading can add to the send buffer.

Reply:
Right, I've re-read the datasheet and it looks like you are correct!  In that case, I think your code is okay.